File: deck/seed/code/native/node/network/runtime/http2.ts
// HTTP/2 server runtime for node, over `node:http2`.
//
// TWO WAYS IN, and node has a constructor for each:
//
//   secure   `createSecureServer`, TLS with ALPN advertising h2. The only way a browser speaks HTTP/2.
//   clear    `createServer`, h2c with prior knowledge: no ALPN, no upgrade dance.
//
// node is the one backend where HTTP/2 was already reachable, and it was reachable as a mill macro that nothing
// imported and that two empty files belonged to. This is the same surface the other three have.
//
// Reached only through the public network/http2 API.
import * as http2 from 'node:http2'

type SeedRequest = {
  method: string
  url: string
  path: string
  query: string
  headers: Record<string, string>
  body: string
  dock: unknown
}

type SeedResponse = {
  status?: number
  headers?: Array<{ name: string; value: string }>
  body?: string
}

type Handler = (request: SeedRequest) => SeedResponse | Promise<SeedResponse>

type Running = { server: http2.Http2Server | undefined; port: number }

// RFC 9113 forbids these; a response carrying one is malformed and node throws on it
const CONNECTION_HEADERS = new Set([
  'connection',
  'keep-alive',
  'proxy-connection',
  'transfer-encoding',
  'upgrade',
])

const http2Runtime = {
  serve: (
    port: number,
    host: string,
    handler: Handler,
    secure: boolean,
    certificate: string,
    key: string,
  ): void => {
    void http2Runtime.start(port, host, handler, secure, certificate, key)
  },

  start: async (
    port: number,
    host: string,
    handler: Handler,
    secure: boolean,
    certificate: string,
    key: string,
  ): Promise<Running> => {
    const scheme = secure ? 'https' : 'http'
    const onStream = (
      stream: http2.ServerHttp2Stream,
      headers: http2.IncomingHttpHeaders & http2.IncomingHttpStatusHeader,
    ): void => {
      void answer(stream, headers, handler, scheme)
    }

    const server = secure
      ? http2.createSecureServer({ cert: certificate, key, allowHTTP1: true })
      : http2.createServer()

    server.on('stream', onStream)

    await new Promise<void>(ready => {
      server.listen(port, host, () => ready())
    })

    return { server, port }
  },

  stop: async (running: Running): Promise<void> => {
    const server = running.server
    running.server = undefined

    if (!server) {
      return
    }

    await new Promise<void>(done => {
      server.close(() => done())
    })
  },
}

async function answer(
  stream: http2.ServerHttp2Stream,
  headers: http2.IncomingHttpHeaders,
  handler: Handler,
  scheme: string,
): Promise<void> {
  const target = String(headers[':path'] ?? '/')
  const split = target.indexOf('?')
  const path = split >= 0 ? target.slice(0, split) : target

  const flat: Record<string, string> = {}

  for (const [name, value] of Object.entries(headers)) {
    flat[name.toLowerCase()] = Array.isArray(value)
      ? value.join(', ')
      : String(value ?? '')
  }

  // the pseudo-headers node already gives us, plus the stream id it does expose (unlike the other three)
  flat[':scheme'] = String(headers[':scheme'] ?? scheme)
  flat[':authority'] = String(headers[':authority'] ?? flat['host'] ?? '')
  flat[':path'] = target
  flat['x-term-stream'] = String(stream.id ?? 0)

  const body = await new Promise<string>(done => {
    const parts: Buffer[] = []
    stream.on('data', (chunk: Buffer) => parts.push(chunk))
    stream.on('end', () => done(Buffer.concat(parts).toString('utf8')))
    stream.on('error', () => done(''))
  })

  let answered: SeedResponse

  try {
    answered = await handler({
      method: String(headers[':method'] ?? 'GET'),
      url: target,
      path,
      query: split >= 0 ? target.slice(split + 1) : '',
      headers: flat,
      body,
      dock: stream,
    })
  } catch {
    answered = { status: 500, body: '' }
  }

  const out: Record<string, string | string[]> = {
    ':status': String(answered.status ?? 200),
  }

  for (const header of answered.headers ?? []) {
    if (CONNECTION_HEADERS.has(header.name.toLowerCase())) {
      continue
    }

    const already = out[header.name]

    // repeats survive, which is what several set-cookie headers needs
    if (already === undefined) {
      out[header.name] = header.value
    } else if (Array.isArray(already)) {
      already.push(header.value)
    } else {
      out[header.name] = [already, header.value]
    }
  }

  try {
    stream.respond(out)
    stream.end(answered.body ?? '')
  } catch {
    stream.close()
  }
}

File: deck/seed/code/native/swift/network/runtime/http2.swift
// HTTP/2 server runtime for the swift target, over HummingbirdHTTP2 (swift-nio-http2 underneath).
//
// TWO WAYS IN, and Hummingbird has a builder for each:
//
//   secure   `HTTPServerBuilder.http2Upgrade(tlsConfiguration:)`, TLS with ALPN advertising h2, falling back to
//            HTTP/1.1 for a client that cannot do h2. This is the only way a browser speaks HTTP/2.
//   clear    `HTTPServerBuilder.plaintextHTTP2()`, h2c with prior knowledge: no ALPN, no upgrade dance.
//
// Adding this product is also what gives the swift target TLS AT ALL: h2 is negotiated by ALPN, so the HTTP/2
// package IS the TLS package (NIOSSL). network/runtime/server.swift uses it too now, and no longer serves
// plaintext while apologising for it.
//
// Reached only through the public network/http2 API. See note/term/stdlib/native-async-file-and-server.md.
import Foundation
import HTTPTypes
import Hummingbird
import HummingbirdCore
import HummingbirdHTTP2
import NIOCore
import NIOPosix
import NIOSSL

enum http2 {
  final class Running: @unchecked Sendable {
    var stop: (() -> Void)?
    let port: Int
    init(stop: (() -> Void)?, port: Int) {
      self.stop = stop
      self.port = port
    }
  }

  // the handler is boxed because Hummingbird runs requests on a concurrent executor and a Term task value is not
  // Sendable: the same single-thread constraint the rust target has, for the same reason
  private final class Held: @unchecked Sendable {
    let call: (Request) -> Response
    init(_ call: @escaping (Request) -> Response) {
      self.call = call
    }
  }

  static func serve(
    _ port: Int,
    _ host: String,
    _ handler: @escaping (Request) -> Response,
    _ secure: Bool,
    _ certificate: String,
    _ key: String
  ) {
    let waiting = DispatchSemaphore(value: 0)

    Task {
      await run(port, host, Held(handler), secure, certificate, key)
      waiting.signal()
    }

    waiting.wait()
  }

  static func start(
    _ port: Int,
    _ host: String,
    _ handler: @escaping (Request) -> Response,
    _ secure: Bool,
    _ certificate: String,
    _ key: String
  ) async -> Running {
    let running = Running(stop: nil, port: port)
    let held = Held(handler)
    let task = Task {
      await run(port, host, held, secure, certificate, key)
    }

    running.stop = { task.cancel() }

    // let the listener bind, so the port is taken by the time this answers
    try? await Task.sleep(nanoseconds: 50_000_000)

    return running
  }

  static func stop(_ server: Running) async {
    server.stop?()
    server.stop = nil
  }

  private static func run(
    _ port: Int,
    _ host: String,
    _ handler: Held,
    _ secure: Bool,
    _ certificate: String,
    _ key: String
  ) async {
    let router = Router()
    let every: [HTTPRequest.Method] = [
      .get, .post, .put, .patch, .delete, .head, .options,
    ]

    // ONE catch-all route per method: the Term API routes inside the handler, so Hummingbird's router is not
    // used for dispatch. Two routers in one stack is how a request is matched twice and answered once.
    for method in every {
      router.on("/**", method: method) {
        (request: Hummingbird.Request, _: BasicRequestContext) async throws
          -> Hummingbird.Response in
        try await answer(request, handler, secure ? "https" : "http")
      }
    }

    let builder: HTTPServerBuilder

    if secure {
      guard let tls = tlsConfiguration(certificate, key) else {
        FileHandle.standardError.write(
          Data("http2 tls: certificate or key did not parse\n".utf8)
        )

        return
      }

      do {
        builder = try .http2Upgrade(tlsConfiguration: tls)
      } catch {
        FileHandle.standardError.write(Data("http2 tls: \(error)\n".utf8))

        return
      }
    } else {
      builder = .plaintextHTTP2()
    }

    let app = Application(
      router: router,
      server: builder,
      configuration: ApplicationConfiguration(
        address: .hostname(host, port: port)
      )
    )

    do {
      try await app.runService()
    } catch {
      FileHandle.standardError.write(Data("http2: \(error)\n".utf8))
    }
  }

  // a NIOSSL configuration that ADVERTISES h2 over ALPN, which is the whole difference from the HTTP/1.1 one
  static func tlsConfiguration(
    _ certificate: String,
    _ key: String
  ) -> TLSConfiguration? {
    do {
      let chain = try NIOSSLCertificate.fromPEMBytes(Array(certificate.utf8))
        .map { NIOSSLCertificateSource.certificate($0) }
      let private_ = try NIOSSLPrivateKey(bytes: Array(key.utf8), format: .pem)
      var tls = TLSConfiguration.makeServerConfiguration(
        certificateChain: chain,
        privateKey: .privateKey(private_)
      )
      tls.applicationProtocols = ["h2", "http/1.1"]

      return tls
    } catch {
      return nil
    }
  }

  private static func answer(
    _ request: Hummingbird.Request,
    _ handler: Held,
    _ scheme: String
  ) async throws -> Hummingbird.Response {
    let headers = SeedMap<String, String>([:])

    for field in request.headers {
      headers.data[field.name.canonicalName] = field.value
    }

    var whole = ByteBuffer()

    for try await chunk in request.body {
      var chunk = chunk
      whole.writeBuffer(&chunk)
    }

    let target = request.uri.string
    let path = request.uri.path

    // the pseudo-headers, in the SAME map the HTTP/1.1 server fills, which is what lets one handler serve both
    headers.data[":method"] = request.method.rawValue
    headers.data[":scheme"] = scheme
    headers.data[":authority"] =
      request.uri.host ?? headers.data["host"] ?? ""
    headers.data[":path"] = path
    // Hummingbird does not surface the stream id to a responder, so it is reported as 0 rather than invented
    headers.data["x-term-stream"] = "0"

    let answered = handler.call(
      Request(
        method: request.method.rawValue,
        url: target,
        path: path,
        query: request.uri.query ?? "",
        headers: headers,
        body: String(buffer: whole),
        dock: 0
      )
    )

    var out = HTTPFields()

    for header in answered.headers.data {
      // RFC 9113 forbids the connection-specific fields; a response carrying one is malformed
      if isConnectionHeader(header.name) {
        continue
      }

      if let name = HTTPField.Name(header.name) {
        out.append(HTTPField(name: name, value: header.value))
      }
    }

    return Hummingbird.Response(
      status: HTTPResponse.Status(code: answered.status),
      headers: out,
      body: ResponseBody(byteBuffer: ByteBuffer(string: answered.body))
    )
  }

  private static func isConnectionHeader(_ name: String) -> Bool {
    switch name.lowercased() {
    case "connection", "keep-alive", "proxy-connection", "transfer-encoding",
      "upgrade":
      return true
    default:
      return false
    }
  }
}
